Add ctor of MIPS instruction from string (#589)

pukhovv opened the pull request.

pukhovv committed:
Add ctor of MIPS instruction from string (#589)

Additionally:
New utility method in class BaseMIPSInstr:
    MapType::iterator find_entry(MapType isaMap, std::string_view name);
New unit tests:
    TEST_CASE( "sllv") (#530)
    TEST_CASE( "MIPS32_instr: Divmult") ( using new ctor)

pavelkryukov commented on simulator/mips/mips_instr.h:
```diff
@@ -161,6 +161,7 @@ class BaseMIPSInstr
         static const MapType isaMapRI;
         static const MapType isaMapIJ;
         static const MapType isaMapMIPS32;
+        typename MapType::iterator find_entry( MapType isaMap, std::string_view name);
```
While calling the function, you copy the `std::map` completely. Consider passing the constant reference instead.

pavelkryukov replied:
The function should return `MapType::const_iterator` as we do not intend to change the maps, and the function should have `const` qualifier as we do not change the object itself. So the signature would be:
```c++
typename MapType::const_iterator find_entry( MapType isaMap, std::srting_view name) const;
```

Also, a typedef would be useful here.

pavelkryukov commented on simulator/mips/mips_instr.cpp:
```diff
@@ -294,6 +294,44 @@ BaseMIPSInstr<RegisterUInt>::BaseMIPSInstr( MIPSVersion version, uint32 bytes, A
+        it =  find_entry( isaMapIJ,     str_opcode);
+    if ( it == isaMapIJ.end())
+    {
+        std::ostringstream oss;
```
The block is duplicated from decoding ctor, please add a separate method

pukhovv replied:
This block is similar, but uses stringed opcode. Should I add another method for the class (opcode and str_opcode variants)?

pavelkryukov replied:
Ok, I see the difference now. Let's leave it as is.

pavelkryukov commented on simulator/mips/mips_instr.cpp:
```diff
@@ -294,6 +294,44 @@ BaseMIPSInstr<RegisterUInt>::BaseMIPSInstr( MIPSVersion version, uint32 bytes, A
+        if ( it->second.name == name)
+            return it;
+    }
+    return it;
```
It's a liitle but unobvious what is returned here. Explicit return of `map.end()` might make code clearer.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -9,6 +9,15 @@ TEST_CASE( "MIPS32_instr_init: Process_Wrong_Args_Of_Constr")
 }
 
+TEST_CASE( "sllv")
+{
```
You test only R-instructions. Could you please add some tests for I, J, RI and other instruction types?

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -9,6 +9,15 @@ TEST_CASE( "MIPS32_instr_init: Process_Wrong_Args_Of_Constr")
+    MIPS32Instr instr("sllv");
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0, 1);
+    instr.execute();
```
Could you please add a test case for non-existing instruction?
It should throw an exception if `execute()` is called.

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
String-ctor fixes and new test features(#589)

Fix types of supporting function.
Add get_v_imm and set_v_imm for unit tests.
Add sllv, mul, addi, teqi tests.
Fix nop bug for ctor, add nop test for new ctor.
Fix some other things from review.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -9,15 +9,59 @@ TEST_CASE( "MIPS32_instr_init: Process_Wrong_Args_Of_Constr")
-    MIPS32Instr instr("sllv");
+    MIPS32Instr instr( "invalid_instruction");
+    bool isvalid = true;
+    try
```
CATCH2 library has an alias for that:
```c++
CATCH_THROWS_AS( instr.execute(), UnknownMIPSInstruction)
```

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Fix invalid_instruction test (#589)
